### PR TITLE
AArch64: Use add/sub shifted register instruction if possible

### PR DESCRIPTION
**Pull request by Akira1Saitoh**
Use add (shifted register) instruction if add operation can be
encoded into this instruction.
Similarly, use sub (shifted register) instruction if sub operation
can be encoded into this instruction.

Signed-off-by: Akira Saitoh <saiaki@jp.ibm.com>

### CODE REVIEW

File: compiler/aarch64/codegen/BinaryEvaluator.cpp [hunk 1]
```diff
@@ -179,6 +179,71 @@ generateMaddOrMsub(TR::Node *node, TR::Node *mulNode, TR::Node *anotherNode, TR:
...
+      {
+      return NULL;
+      }
+
+   const bool is64bit = node->getDataType().isInt64();
+   if ((shiftValue <= 0) || (shiftValue > (is64bit ? 63 : 31)))
+      {
+      return NULL;
+      }
+   TR::Register *treg = cg->allocateRegister();
```
Comment (knn-k) [inline compiler/aarch64/codegen/BinaryEvaluator.cpp:235]: Is it possible to check the reference count of source1Node and source2Node here, to reuse s1reg or s2reg as treg?

File: compiler/aarch64/codegen/BinaryEvaluator.cpp [hunk 2]
```diff
@@ -179,6 +179,71 @@ generateMaddOrMsub(TR::Node *node, TR::Node *mulNode, TR::Node *anotherNode, TR:
...
+   TR::Register *s1reg = cg->evaluate(source1Node);
+   TR::Register *s2reg = cg->evaluate(source2Node);
+   TR::ARM64ShiftCode code = (shiftNode->getOpCode().isLeftShift() ? TR::SH_LSL : (shiftNode->getOpCode().isShiftLogical() ? TR::SH_LSR : TR::SH_ASR));
+
+   generateTrg1Src2ShiftedInstruction(cg, op, node, treg, s1reg, s2reg, code, shiftValue);
+   node->setRegister(treg);
+   cg->recursivelyDecReferenceCount(shiftNode);
+   cg->decReferenceCount(source1Node);
```
Comment (knn-k) [inline compiler/aarch64/codegen/BinaryEvaluator.cpp:243]: What about the reference count of source2Node?

Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/BinaryEvaluator.cpp:243]: source2Node is a firstChild of shiftNode, and we call `recursivelyDecReferenceCount` on shiftNode at line 242, so the reference count of source2Node is properly decremented.

File: compiler/aarch64/codegen/BinaryEvaluator.cpp (hunk 1 above)
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/BinaryEvaluator.cpp:235]: Thanks for the suggestion. Will work on it.

Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/BinaryEvaluator.cpp]: force-pushed the updated commit

**Comment by knn-k** [comment]
Jenkins build aarch64

**Comment by knn-k** [comment]
Jenkins build aarch64

**Review by knn-k — APPROVED** [review]

**Pull request merged**